File: 2021/07/07.py
from statistics import mean, median

def preprocessing(input): 
    return [int(item) for item in input.split(',')]

def solver(crabs):
    med_crabs  = int(median(crabs))
    med_fuel   = 0
    mean_crabs = int(mean(crabs))
    mean_fuel  = 0
    
    for crab in crabs:
        med_fuel  += abs(crab - med_crabs)
        distance   = abs(crab - mean_crabs)
        mean_fuel += distance * (distance + 1) // 2
        
    yield med_fuel
    yield mean_fuel

File: 2021/07/07.rs
aoc::main!();

fn preprocessing(input: &str) -> Vec<usize> {
    input
    .split(',')
    .map(|n| n.parse().unwrap())
    .collect()
}

fn part_1(crabs: Vec<usize>) -> usize {
    crabs
    .iter()
    .map(|crab| med(crabs.clone()).abs_diff(*crab))
    .sum()
}

fn part_2(crabs: Vec<usize>) -> usize {
    crabs
    .iter()
    .map(|crab| mean(crabs.clone()).abs_diff(*crab))
    .map(|dist| dist * (dist + 1) / 2)
    .sum()
}


fn med(numbers: Vec<usize>) -> usize {
    *numbers
    .iter()
    .sorted()
    .nth(numbers.len() / 2)
    .unwrap()
}

fn mean(numbers: Vec<usize>) -> usize {
    numbers
    .iter()
    .sum::<usize>()
    /
    numbers.len()
}
